### ISSUE TITLE
Performance tests for v6 |  #tf-v6-perf-tests

### ISSUE THREAD

**Issue by andreasohlund** [Particular/EndToEnd#3]
Taskforce: William (Lead) / Igal / Greg / Ramon / Dennis / Hadi @WilliamBZA @hmemcpy @gbiellem @ramonsmits @dvdstelt @HEskandari 

Discussion: [#tf-v6-perf-tests](https://particularsoftware.slack.com/messages/tf-v6-perf-tests/)

We need to run end to end performance tests for v6 transports and persistence for all the supported transport transaction modes. 

I have a hunch that we can reuse the code and infra from the outbox tests?

## Plan of Attack

Action plan can be found [here](https://github.com/Particular/EndToEnd/blob/docs/docs/action-plan.md)

**Comment by mauroservienti** [comment]
@dvdstelt is working on something that is related to this as well

**Comment by WilliamBZA** [comment]
I'd like to participate in this, and I suspect @ramonsmits might like to as well.

**Comment by WojcikMike** [comment]
Count me in

**Comment by andreasohlund** [comment]
Given the potential overlap with #2 how should we play this?

**Comment by WojcikMike** [comment]
@andreasohlund so I assume you prefer to have 2 separate TF ideally without people overlap?

**Comment by andreasohlund** [comment]
Or we start with either #2 or #3 to avoid overlap since both will build on the outbox tests?

**Comment by WojcikMike** [comment]
sounds good to me. Pick one that is more important and one can start with it

**Comment by andreasohlund** [comment]
I'd vote for performance first?

If we agree can you sync with the TF of #2 to decide if they want to join here instead @WojcikMike ?

**Comment by gbiellem** [comment]
@WojcikMike 

>  @gbiellem @johnsimons are you interested and have time please comment on #3

Sure count me in.

**Comment by ramonsmits** [comment]
Would be willing to participate in this as I have created numerous stress tests for SQL, Rabbit and MSMQ.

**Comment by andreasohlund** [comment]
So we have @WojcikMike @gbiellem @ramonsmits @WilliamBZA and potentially @dvdstelt interested in this one. Let's get it going, will take the lead? When can we have a preliminary POA ready?

**Comment by andreasohlund** [comment]
@dannycohen can you chime in with the data want to get out of this from a "this is why you should use v6" perspective?

**Comment by WilliamBZA** [comment]
### Items to discuss and work on

 - [ ] Identify requirement outcomes
   - Do we plan to use the results to lure prospects?
   - Or do we want to have a performance baseline and test for regression in performance?
 - [ ] Copy OutboxPerformanceTests to EndToEnd Repo
 - [ ] Decide on the metrics to be measured  
   - Sustained message throughput?
   - Start-up time?
   - Variance on high-error amounts?
   - Differences with auditing on/off?
 - [ ] Decide on the measurement technique (is Metrics.Net sufficient?)
 - [ ] Refactor test harness to be more flexible (taking the permutations of Transports + Persistence + Outbox on/off + Auditing on/off already gives 100 different cases)
   - Maybe scan for `ITestCase` where each `ITestCase` defines the transport and persistence to use, as well as cleans up after itself and reports measurement data to the harness
 - [ ] Implement different test cases
 - [ ] Decide how we're going to automate these test runs on the build server
   - Run _all_ tests cases each time?
   - Run on each check-in? (probably an awful idea)
   - Run daily?  
   - How is the build server going to report performance stats? Failing builds?
   - How should we respond to code the introduces performance problems?

@dannycohen @andreasohlund can you please give us your thoughts for the first point in the meantime so we can understand the context of this task and make sure we move in the right direction?

**Comment by ramonsmits** [comment]
Additional metrics:

- Time to first message

**Comment by andreasohlund** [comment]
Does it make sense to check CPU, Memory usage, thread count and IO utilization as well?

**Comment by ramonsmits** [comment]
Measuring each metric is always in correlation with the profile. Auditing for example, is part of that profile and the same for things like encryption but also environment things like number of cores, CPU specs.

If we don't re-use the same machine between runs in Azure / EWS then this has an impact on results as there can be CPU difference even though you choose the same instance type.

**Comment by WilliamBZA** [comment]
@andreasohlund I think definitely, yes. Thats was something that @ramonsmits brought up while we were playing around with our tests. Knowing what the environment was, and how it was responding during the tests would have allowed us to gain insights and make educated guesses about bottle-necks.

**Comment by ramonsmits** [comment]
I was also considering 'environment' in the profiles. With that I mean, Azure, AWS, AWS + reserved IOPS and/or On-premise as there behave incredibly different in regards to IO.

The first step would be to have an environment to run the tests in a replicable manner. We should test if we get similar results when the tests are run on a new VM.

**Comment by ramonsmits** [comment]
Also, how do we aggregate results so that we can see historical results? Seems to me we probably are not the first in solving this run a benchmark as part of a build, store its results by pull request/branch/commit/versionnr. and being able to view these resulst, see trends, compare by profile, etc.

**Comment by andreasohlund** [comment]
In the past we've used TeamCity custom stats https://confluence.jetbrains.com/display/TCD8/Custom+Chart but that was quite clunky to set up and work with so I suggest we look for other options if this is a must have

**Comment by ramonsmits** [comment]
We can potentially use https://github.com/nunit/docs/wiki/TestCaseData to run all benchmarks as regular unit tests?

If we use nunit as then we must make sure that the tests are not run in parallel as that would have a negative impact on results.

**Comment by ramonsmits** [comment]
Would also be nice to maybe add a couple of 'native' tests to get some baselines on performance of native Rabbit, MSMQ ASB, ASQ so we can compare how much is lost due to the middleware.

For example, when not using any of the reliability features and doing a basic Bus.Send(destination) would give us a nice indication of this.

**Comment by andreasohlund** [comment]
>We can potentially use https://github.com/nunit/docs/wiki/TestCaseData to run all benchmarks as regular unit tests?

Would this be suitable for long running tests like we plan in #2 or is that out of scope?

**Comment by WilliamBZA** [comment]
Also should we measure how much concurrency and other settings affects the results? Or should each `TestCase` have it's pre-measured and pre-configured expected best-case settings?

**Comment by dvdstelt** [comment]
@ramonsmits Measuring perf without reliability features would be nice so we get better insight ourselves and can optimize from there. We should be careful with communicating this outside. We will make the tests public, but how about perf tests & results? Because people who lack performance in their own apps, might think they need to drop transactions and outbox and what not, for a little performance gain. This will hurt the in the long run.

So the question, are we doing perf tests for internal use, or are we going public with the numbers as well?

**Comment by dannycohen** [comment]
@WilliamBZA - Regarding "Identify requirement outcomes" (https://github.com/Particular/EndToEnd/issues/3#issuecomment-199233858) - here are my thoughts - potentially as a "output discussion starter" - 

Looking at this issue from the perspective of attempting to define a V6 "Customer Value proposition" (https://github.com/Particular/PlatformDevelopment/issues/660) - 

1. The main dependency on performance tests is for validating (or invalidating) some of the assumptions and benefits V6 talking points 
* Specifically, from the points defined in https://github.com/Particular/PlatformDevelopment/issues/660: 
   1. Dramatic throughput enhancements
   2. Significant resource utilization improvements
* So that can be the source from which we derive the required outcomes


## Assumption 1. Dramatic throughput enhancements

1. **Question: whats is the change in message throughput (measured in # of messages/sec) between NSB V5 and NSB V6 ?** 
* Should we also include NSB V4 since it is a widely used version ?
* Measurement assumptions:
   * Measure "clean" throughput - without any significant IO/CPU workload on the endpoints. This is potentially less realistic but focuses and highlights NSB evolution which is the point of this measurement
   * Replicate measurements for all NSB out-of-the-box supported transports 
   * Hardware / hosting: I believe a medium-large Cloud VM hosting would be sufficient. The absolute throughput numbers should be large and impressive enough while the "small print hardware requirements" should not be daunting. Main objective here is to highlight difference between NSB versions (throughput differences between hardware profile is a different - and valuable - question)  
* Potential outcome of this investigation that can be used for various purposes:

![image](https://cloud.githubusercontent.com/assets/3889023/13930406/91d66c98-efa7-11e5-9422-73a284e4fae1.png)
   
## Assumption 2. Significant resource utilization improvements

1. The goal of this specific question is to evaluate the cost changes of NSB V6 support for multiple endpoints being hosted in the same process.
* Therefore, the questions are:
   * What is the RAM footprint of an NSB Endpoint in V6 ? (assuming multiple endpoints are hosted in the same process)
   * How many NSB V6 endpoints can we co-host on the same machine (compared to NSB pre-V6) ? 
   
Thoughts ?

**Comment by ramonsmits** [comment]
The following gives a preliminary list of options regarding environments, configuration and test permutations.

Let's discuss tomorrow which is these are part of this scope and prioritize.


## Environments

- [ ] Azure
- [ ] AWS (https://aws.amazon.com/ec2/instance-types/)

RavenDB

- [ ] Local
- [ ] Remote

SQL Server

- [ ] Local
- [ ] Remote
- [ ] SQL Azure


Oracle

- [ ] Local
- [ ] Remote

RabbitMQ

- [ ] Local
- [ ] Remote

## Configurations

Logging

- [ ] Lowest level

NServiceBus

- [ ] 4.x
- [ ] 5.x
- [ ] 6.x

Transports:

- [ ] MSMQ
- [ ] RabbitMQ
- SqlServer
 - [ ] 1.x
 - [ ] 2.x
 - [ ] 3.x
- [ ] ASQ
- [ ] ASB

Persistence Multiple major versions per core major version e.g. NServiceBus V5 - SQL Transport 1.x &  NServiceBus V5 - SQL Transport 2.x

- [ ] NHibernate 
- [ ] RavenDB
- [ ] InMemory

Serialization

- [ ] XML
- [ ] JSON
- [ ] Binary??

Auditing

Encryption

Message size:

- [ ] Tiny (Headers only, empty message)
- [ ] Small (1kb)
- [ ] Medium (10kb)
- [ ] Large (64kb)

Outbox

Transactional modes

- [ ] DTC
- [ ] No transactions
- [ ] Native

Concurrency

- [ ] 1
- [ ] Environment CPU count?
- [ ] 4 * Environment CPU count?

Operations

- [ ] Publish
- [ ] Send (internal routing)
- [ ] Send with destination
- [ ] Durable messages (Express?)
- [ ] Defer ?

Sagas

- [ ] Concurrency

Metrics

- [ ] Time / Duration
- [ ] Total number of messages
- [ ] Sustained message throughput
- [ ] Peak message throughput
- [ ] Time to first message
- [ ] Concurrency
- [ ] FLR count
- [ ] Message age
- System & Endpoint process utilization
  - [ ] CPU Utilization
  - [ ] Disk Utilization
  - [ ] Memory Utilization
  - [ ] Network Utilization

Measurement

- [ ] On time (in 1 minute we were able to process X messages)
- [ ] On message count (it took X seconds to process 10,000 messages)

Hosting

- [ ] AppDomain
- [ ] External process

## Out of scope?

- Timeouts
- Containers
- Distributor
- Gateway

**Comment by ramonsmits** [comment]
## Comparing resource utilization

In order to test if V6 has a lower resource utilization, we should do processing in the same rate between both versions. This requires limiting on either the sender and/or receiver.

As V6 dropped limiting it seems our only option is to rate limit sends that trigger the whole benchmark flow.

Now we are able to compare CPU, RAM, etc. utilization for the same sustained throughput.

For example: At 200 message per second the RAM, CPU, IO utilization is X for V5 and Y for V6.

**Comment by andreasohlund** [comment]
>Out of scope?

>Timeouts
Containers
Distributor
Gateway

Do we include the client side distribution? ( @SzymonPobiega @janovesk there seems to be an outstanding task to perf test that in https://github.com/Particular/PlatformDevelopment/issues/133 )

**Comment by ramonsmits** [comment]
https://github.com/Particular/ProductionTests/tree/master/src/PerformanceTests

Things from this project that we potentially can use:

- Has a common shared project to share source code between major versions
- Is using NUnit as the test runner
- Creates an app domain per test
 - Might not work with tracking all metrics in regards to utilization.
 - Could also create a new process
- High isolation for improving utilization metrics
- Has an endpoint generator
- Each permutation has its own artifacts folder (package versions, etc.)
- Outputs results as NUnit
 - Each test is tracked by TeamCity
![image](https://cloud.githubusercontent.com/assets/601206/13956721/73d2123c-f052-11e5-8cf2-3392beb478eb.png)
- Still needs a separate view to display all captures metrics per test.
- By using NUnit its hard to scale out the test runs across a large number of agents / virtual machines
 - We might be able to spread tests by using different build configurations

**Comment by ramonsmits** [comment]
How are we going to visualize captured metrics? These is the native NUnit but that will not automatically collect metrics. Using Metrics.net helps in capturing metrics but not comparing them.

Maybe any these can be used:

- https://github.com/PerfDotNet/BenchmarkDotNet
- https://petabridge.com/blog/introduction-to-nbench/

**Comment by ramonsmits** [comment]
@andreasohlund We could benchmark the client side distribution but that would then compare against the V5 distributor. Isn't that comparing apples to oranges? I think we currently should focus on validating **Assumption 1. Dramatic throughput enhancements**

**Comment by ramonsmits** [comment]
In order to remove randomization due to concurrency we should probably rely on generating deterministic message ID's, then this value can be used as the pseudo randomization seed. This could even be combined with maybe FLR count

**Comment by SzymonPobiega** [comment]
@andreasohlund @ramonsmits yes, please include the client-side distribution (only for MSMQ) to the test matrix. I don't see a reason to run these tests separately from this effort.

**Comment by dvdstelt** [comment]
As mentioned yesterday, we should test "publish" and "send" and such in isolation, but also test a single handler for multiple actions, so that we verify a real test.
- incoming message
- publish message
- send message

or
- incoming message into existing saga & publish and/or send messages
- incoming message into non-existing saga & publish and/or send messages

And other variations we can think of. We had this with ServiceControl before with a large number of incoming messages, where retry of messages failed because it was busy with incoming messages.

**Comment by dvdstelt** [comment]
@ramonsmits You think we should run tests using NUnit? I don't think it automatically runs tests in parallel, which would have a negative effect on performance. And we can't say anything about ordering of tests, especially if we add tests later.

The TransportCompat has tests set up for all transports and we might be able to use code from it as well.

**Comment by WojcikMike** [comment]
## Meeting 23rd of March 2016
Attendees: Igal, Dennis, Ramon, Michał, Danny, Andreas
- We should cut scope to do MVP as to have some test before Alpha/Beta release of NSB
- Testing performance of:
 - Publish performance
 - Saga performance
 - Message Throughput
 - Databus usage
- Create production like example, for example when handling a message:
 - Calling DB
 - Calling WebService?
- On the call we made initial agreement on what will be tested (check Ramon's comment with Environment) 
- Set up of the environment:
 - We should agree on version of Raven, Rabbit, MS SQL Server, Oracle etc.
 - In the end we will do permutation of the versions, we need to think how to do it
- We should create a list of assumptions for the tests

**Comment by gbiellem** [comment]
>   We should agree on version of Raven, Rabbit, MS SQL Server, Oracle etc.

If we want multiple versions of SQL and Oracle I'll need to build some new agent images or we have the DB servers remote from the test machine which is more real world.  Running several copies of SQL and Oracle in the same VM image isn't a good option.

Once the list is compiled let me know and I can provision agent images or standalone DB servers

**Comment by ramonsmits** [comment]
- DTC Performance
 - if we check DTC performance we must use remote DB and queue to get production like results @andreasohlund 
- Testing persistence performance is hard
 - Seed data to generate a set that is sufficient in size to introduce latency
- In the variables we need to include the infrastructural dependency version numbers
 - RavenDB v3.0.x, MSSQL 2014, RabbitMQ 3.6.x
- Each permutation can be tested in isolation thus on a separate set of machines. This might help in running test runs in parallel reducing time to run all permutations
- What tools can be use to automate creation of enviroments
 - TeamCity
 - Docker
 - Chef
 - Puppet
 - Ansible
 - SaltStack
 - Vagrant
 - TeamCity Agents
 - EC2
 - Azure

**Comment by ramonsmits** [comment]
Benchmarks on the web:

RestBus.Benchmarks
- Compares RestBus to MassTransit and NServiceBus
- https://github.com/tenor/RestBus.Benchmarks

**Comment by gbiellem** [comment]
@ramonsmits 

> What tools can be use to automate creation of environments 

I looked into this a while ago when @andreasohlund  and I were discussing setting up long running tests.  From that investigation I got pretty excited about AWS Cloud Formation scripts. They allow you to build up template designs of multi machine environments, which you can then kick off one or more instances of via script or through the Amazon Web UI

Basically you build a library of VM images and the templates details what machines to build from what images and also how they are networked together etc,  You can also nest templates.  One of the templates I looked at built a complete windows domain complete with multiple app servers and a failover SQL cluster.

**Comment by gbiellem** [comment]
> Each permutation can be tested in isolation thus on a separate set of machines.

One other thing to consider is the limitations put in place by the cloud providers.  For instance by default Azure limits the number of running Virtual CPU per subscription to 20.  Amazon has a similar limit based on IOPS.  Depending on how many variations we want to run concurrently this may be a constraint.  

We can ask for increases in the limits but we have to propose a figure, they won't just remove them.

**Comment by ramonsmits** [comment]
## Terminology

The permutation meeting agreed on a way on how to run the most value benchmark results and defined some terminology:

> Variable

An item that  has multiple values. Examples are:

- Transport
- Persistence
- DTC
- Serialization
- NServiceBus verion

In code this could be an `enum`

> Value

The possible values for a variable. Examples are:

- On, Off
- NServiceBus.RabbitMQ, NServiceBus.MSMQ
- 1,2,3,4,5
- Tiny, Small, Medium, Large
- 0%, 50%, 80%
- 5.2.17, 6.0.0

> Mode

The configuration of a transport. Examples are:

- Local, Remote
- Default, No ACK
- Connection Encryption, Large connection pool

> Environment

The identification of the environment. Examples are:

- On-Premise
- Azure
- Amazon

> Instance type

The type of machine on which the benchmark is run. Example are:

- A0
- D2
- G5
- t2.nano
- c4.2xlarge
- i2.2xlarge


> Category

A specific set of variable values. The following examples do not mention the defaults, only the values that influence the number of permutations

PAAS vs IAAS

- Mode: SQL Azure, SQL Remote
- Version: 5.2.x, 6.0.x

Outbox vs DTC

- Version: 5.2.x, 6.0.x
- DTC: On, Off
- Outbox: On, Off

Publish vs Send

- Version: 5.2.x, 6.0.x
- Transport: RabbitMQ, MSMQ, SQL, ASB, ASQ
- Persistence: InMemory, NHibernate, RavenDB

> Permutation

A specific combination of Value, Mode and Environment. This basically identies a specific benchmark configuration.

**Comment by ramonsmits** [comment]
## Benchmark permutations

We have made a sheet where we defined the *categories* and the number or variable values to see how many permutations we have.

![image](https://cloud.githubusercontent.com/assets/152998/13987561/6dcbd34e-f108-11e5-8b9f-5c60ac1a1ff5.png)

(screenshot misses distributor + client side distribution categories)

Google sheet Benchmark permutations:
- https://docs.google.com/a/nservicebus.com/spreadsheets/d/1avUW8Y5gpcPqTxIBaq7X5OXXaE4lDU0e0ZA9FDFNygs/edit?usp=sharing

Combining all possible values for the alpha / beta release would result in 24,000 (!!) permutations. The defined catagories results in **just 211** permutations benchmark.

## Run duration of all benchmark permutations

If each permutationwould take 1 minute then **running all permutations would take about 3.5 hours** exluding environment setup / teardown. This should not be an issue to run on a daily or weekly or on-demand basis.

**Comment by ramonsmits** [comment]
## Choice of environment

Based on the current variables it makes the most sense to **run all benchmarks in Azure** due to us supporting Azure Service Bus, Azure Storage Queues and indirectly also Azure SQL.

Running in Amazon EC2 makes sense later especially when comparing to other benchmarks as Amazon EC2 seems to be the defacto standard to use for benchmarking IAAS solutions.

**Comment by SeanFeldman** [comment]
> by default Azure limits the number of running Virtual CPU per subscription to 20

This can be increased through a request with support.

**Comment by gbiellem** [comment]
@SeanFeldman  yep - That's what I meant when i said:

> We can ask for increases in the limits but we have to propose a figure, they won't just remove them.

**Comment by gbiellem** [comment]
> based on the current variables it makes the most sense to run all benchmarks in Azure due to us supporting Azure Service Bus, Azure Storage Queues and indirectly also Azure SQL.

Not sure I agree - There is nothing preventing us use ASQ and ASB from Amazon and the AWS tooling is better for multi machine setups and tear downs .

If we are doing apples to apples comparisions then we are only interested in perf between V5 and V6 not the raw numbers.  it would be interesting to see how much of an affect running the same rest in Azure and AWS has given the huge pipes and low latency between these guys in the US.

**Comment by ramonsmits** [comment]
@gbiellem Using ASB and ASQ from AWS would increase latency (assumption here..) compared to using that in the same data center which would have bad impact on throughput and / or average processing time.

**Comment by andreasohlund** [comment]
Doesn't this highlight that we have to separate the "test administration" infrastructure from the actual running of the tests? (ie we should be able to run things in both AWS and Azure)

Ie: Not run the tests on the build agents

**Comment by ramonsmits** [comment]
@andreasohlund I do agree, in the end we should be able to run tests anywhere.. but for the beta we should focus on one environment.

**Comment by andreasohlund** [comment]
Is not running on the buildagents in scope? (seem like that's going to break down pretty soon as the number of permutations rise)

**Comment by dvdstelt** [comment]
> Using ASB and ASQ from AWS would increase latency (assumption here..)

The difference is very noticeable from East to West USA, so it should be as well from AWS to Azure. And SQL Azure is included in the permutations. We could run all those from AWS, but if we make those numbers public, we're cutting in our own fingers. If Amazon is good at setting up environments, is simply copying them to Azure an option?

**Comment by gbiellem** [comment]
I assumed the build agents would trigger and control the process but not be the machines that run the code.  I thought the process would be similar to the SC/SI/SP installer testing we did before.

In that case the agent built the code and then triggered a deployment step which would setup the required test machines on azure from images, deploy the code to those machines, run the tests and then report back.  Once all machines had reported back the build agent ran another step to tear down those machines.  All of this was done via powershell scripts there was no Azure tooling that really help out here.

As far as AWS vs Azure goes the variation would be in the deployment scripts and the cost.
I agree with @ramonsmits that we should just start with one.  I voted for AWS because the tooling like cloud formation would get it up and running quicker and it's cheaper to run but if If we are interested in peak performance and publishing the raw numbers then @ramonsmits has made the better choice.


  





-

**Comment by danielmarbach** [comment]
@johnsimons raised a question regarding the memory usage of NSB v6. I did a quick run under small load

## Setup

Azure Standard DS4 (8 cores, 28 GB memory)
199820 messages in roughly 2 h 15 minutes
sent in bursts up to 5000 message (randomly generated) in intervals of 30 up to 300 seconds
Sender is using file based routing, routing table is updated every 5 seconds to trigger the file watcher
Receiver side is scaled out
Msmq, Standard Concurrency of 100, XmlSerializer

[Code](https://drive.google.com/open?id=0Bxf5pxyNKtQVNThsSVpfZGNmbHM)

![199820messages](https://cloud.githubusercontent.com/assets/174258/14021758/2ece899a-f1dd-11e5-9808-e62aa1e52c34.png)

Memory Dumps can be found [here](https://drive.google.com/open?id=0Bxf5pxyNKtQVUWtxTnhaXzZRams)

## Sender

![sender cpu](https://cloud.githubusercontent.com/assets/174258/14021898/aefaa34c-f1dd-11e5-866b-c21eb1fa53a1.png)
![sender memory](https://cloud.githubusercontent.com/assets/174258/14021896/aef90fdc-f1dd-11e5-9cdd-7eb3aef097c2.png)
![sender performancecharts](https://cloud.githubusercontent.com/assets/174258/14021897/aefaadf6-f1dd-11e5-8ad1-b260a0945d77.png)
![sender summary](https://cloud.githubusercontent.com/assets/174258/14021899/aeff3114-f1dd-11e5-8509-a0d969946a9c.png)

## Receiver 1

![receiver 1 cpu](https://cloud.githubusercontent.com/assets/174258/14021909/bfd139b0-f1dd-11e5-8816-050c61116cb7.png)
![receiver 1 memory](https://cloud.githubusercontent.com/assets/174258/14021910/bfd49c7c-f1dd-11e5-9543-2181381c8370.png)
![receiver 1 performancecharts](https://cloud.githubusercontent.com/assets/174258/14021911/c00f8c2e-f1dd-11e5-99ca-9fa14b7ea4c0.png)
![receiver 1 summary](https://cloud.githubusercontent.com/assets/174258/14021912/c018e6ac-f1dd-11e5-884d-a150a6175d48.png)


## Receiver 2

![receiver 2performancecharts](https://cloud.githubusercontent.com/assets/174258/14021927/c9d896a6-f1dd-11e5-9bc4-60760d4ca85f.png)
![receiver 2 cpu](https://cloud.githubusercontent.com/assets/174258/14021928/c9d90ec4-f1dd-11e5-881c-04559fb279fd.png)
![receiver 2 memory](https://cloud.githubusercontent.com/assets/174258/14021929/c9dd69ce-f1dd-11e5-8cd9-ffd79d0dffa9.png)
![receiver 2 summary](https://cloud.githubusercontent.com/assets/174258/14021926/c9d72348-f1dd-11e5-98ac-13d222e399fb.png)

**Comment by danielmarbach** [comment]
@johnsimons raised a question regarding the memory usage of NSB v6. I did a quick run under small load

## Setup

Azure Standard DS4 (8 cores, 28 GB memory)
795487 messages in roughly 41 minutes
sent in bursts up to 10000 message (randomly generated) in intervals of 10 up to 30 seconds
Sender is using file based routing, routing table is updated every 5 seconds to trigger the file watcher
Receiver side is scaled out
A really small amount of messages failing
Using 3 senders this time
Msmq, Standard Concurrency of 100, XmlSerializer, InMemoryPersistence, only message handling

[Code](https://drive.google.com/open?id=0Bxf5pxyNKtQVNThsSVpfZGNmbHM)

![795487messages](https://cloud.githubusercontent.com/assets/174258/14024240/9616b540-f1e8-11e5-8458-228800aaad11.png)


All results, screenshots etc. uploaded to [GDrive](https://drive.google.com/drive/u/0/folders/0Bxf5pxyNKtQVQ1pPMHV4UWFLWWs).

## Sender

![sender cpu](https://cloud.githubusercontent.com/assets/174258/14024298/cc732768-f1e8-11e5-9403-9d564f747826.png)
![sender memory](https://cloud.githubusercontent.com/assets/174258/14024301/cc785eb8-f1e8-11e5-9ce8-cdfebb81f748.png)
![sender performancecharts](https://cloud.githubusercontent.com/assets/174258/14024300/cc7803aa-f1e8-11e5-91ad-64ac5e3bf8a3.png)
![sender summary](https://cloud.githubusercontent.com/assets/174258/14024299/cc73f36e-f1e8-11e5-8d0e-74b6768f0526.png)

## Receiver 1

![receiver 1 cpu](https://cloud.githubusercontent.com/assets/174258/14024314/dac3ce6c-f1e8-11e5-87ec-6d6adeea74cf.png)
![receiver 1 memory](https://cloud.githubusercontent.com/assets/174258/14024315/dac44234-f1e8-11e5-98ed-0dc204e5f2b0.png)
![receiver 1 performancecharts](https://cloud.githubusercontent.com/assets/174258/14024317/dac5d982-f1e8-11e5-9d7d-c0defca79464.png)
![receiver 1 summary](https://cloud.githubusercontent.com/assets/174258/14024316/dac53644-f1e8-11e5-9dc5-45787d50cc73.png)

## Receiver 2

![receiver 2 summary](https://cloud.githubusercontent.com/assets/174258/14024326/e1a27e54-f1e8-11e5-9858-03b383b395d4.png)
![receiver 2 cpu](https://cloud.githubusercontent.com/assets/174258/14024328/e247c13e-f1e8-11e5-869a-33c35b7abbbe.png)
![receiver 2 memory](https://cloud.githubusercontent.com/assets/174258/14024329/e2683770-f1e8-11e5-94b2-26920409936b.png)
![receiver 2 performancecharts](https://cloud.githubusercontent.com/assets/174258/14024330/e2869c1a-f1e8-11e5-9742-4f6cbf467a49.png)

**Comment by ramonsmits** [comment]
@danielmarbach What made those graphs?

**Comment by danielmarbach** [comment]
@ramonsmits A drawing fairy tale called [YourKit](https://www.yourkit.com/). Has a few cool options like adding macros which do auto-snapshotting and such.

**Comment by ramonsmits** [comment]
@gbiellem Is going to be creating the machine templates as listed in the issue description.

**Comment by hmemcpy** [comment]
I ran some preliminary tests, launching an existing performance test both as a separate process (Process.Start) and inside an AppDomain (AppDomain.ExecuteAssembly(path)).
Tests were configured to run using 1 thread and limited message concurrency of 10. processing 1000 messages. These are the results:

![image](https://cloud.githubusercontent.com/assets/601206/14065683/61239294-f43c-11e5-9334-a83c2b18f6f9.png)

I noticed that the differences are negligible for the most part, I am starting to collect perf information only after the endpoint had started (so the process/queue creation overhead is not a factor).
In both cases I was able to collect performance counter data.

Note: not sure those numbers are accurate, but the goal was to see any significant perf degradation between running as process/appdomain.

**Comment by ramonsmits** [comment]
@hmemcpy The goal was not to see if there is a performance difference. The goal was to see if all metrics that we want to collect are available on the app domain.

**Comment by andreasohlund** [comment]
Moved this into inprogress

**Comment by ramonsmits** [comment]
## Meeting 2016-03-29

- Did a summary of last week
 - Variables, permutations, categories, research
- Decided on which metrics to collect (see checks in metrics.md)
- Collecting metrics external to not mess with resource utilization of the endpoint
- Collect metrics of all machines to prevent bottlenecks in infrastructure
- Beta launch does not have to be fully automated
 - Can be manually deployed to running environment
 - Tests can maybe be put in a nuget package too to update these
 - In the future a TC agent could kick off the run
- How to cope with configuration between 'dev' box vs benchmark environment

**Comment by WilliamBZA** [comment]
Status update:
Migrated `ProductionTests/PerformanceTests` to `EndToEnd/PerformanceTests` ensuring tests are generated for each released version.


Working next on:
 - adding metrics to generated tests
 - collating metric results
 - Adapting the tests to suit a real-world scenario

**Comment by WilliamBZA** [comment]
Status update:

Action plan has been moved to a markdown file so we can track the changes as we go along.
We have more clarity on the way forward regarding the isolation of test variables
The taskforce is moving towards a more timezone diverse one

**Comment by SeanFeldman** [comment]
> The taskforce is moving towards a more timezone diverse one
Do you need someone from Americas?

**Comment by WilliamBZA** [comment]
@SeanFeldman Thanks for the offer! Not at the moment though - we're still trying to find our way. We will reach out there first once we have need for more dev capacity.

**Comment by WilliamBZA** [comment]
Uploaded sample CSV output files [here](https://drive.google.com/drive/u/1/folders/0B8bTl5PBF8mTTTVNMzFXdkxyWjQ)

**Comment by SeanFeldman** [comment]
Sound good.
Liked the assumption @WilliamBZA :)

**Comment by WilliamBZA** [comment]
Status update:

Test source generation has been completed
Test permutation generation is in progress. Need to still create the bin output for each permutation so it can be run
Output will be recorded directly within splunk to allow for tracking of metrics and history comparisons. The conversion to splunk is in progress
Separation of seeding test data and test data cleanup is being separated from the recording of metrics to prevent contamination of results

**Comment by WilliamBZA** [comment]
Status update:

Completed tasks:
 - Test permutations and variables generate correct binary outputs now
 - All output has been converted to trace outputs for integration into splunk

Ongoing tasks:  
- @hmemcpy to work on the test artifact cleanup
- @ramonsmits  to trigger the processtart code
- @dvdstelt  to include his test cleanup code in the test run after processstart
- @HEskandari to convert trace's to splunk
- @Williambza to start looking at the results to provide feedback for vAlpha

**Comment by ramonsmits** [comment]
Recording:
[![LCTQpCjVt3U](https://i.ytimg.com/vi_webp/LCTQpCjVt3U/mqdefault.webp)](https://youtu.be/LCTQpCjVt3U)

**Comment by WilliamBZA** [comment]
Status update:

* Azure environment is set up and running tests
* Transport tests are running and collecting data
* Persistence tests are in-progress
* Dashboard of stats recorded is being tweaked to represent the required graphs
* Pre-test setup and post-test cleanup is being investigated to ensure the test results aren't skewed
* Azure environment is being changed to support testing distributor model

**Comment by andreasohlund** [comment]
Would it be an option to scope the distributor tests out to a separate task force or is it a must have to gather reasons why users should go to v6? (cc @dannycohen )

**Comment by WilliamBZA** [comment]
@andreasohlund we won't actually be implementing the distributor tests as part of this TF. We'll pause prepping the environment to support it though as that could form part of the separate task.

**Comment by andreasohlund** [comment]
:+1: can you raise a new issue for that and put the consider for v6 label on?

**Comment by gbiellem** [comment]
Azure VMs are now built from images each run so the disk contents are reset. 
This means local cleanup of MSMQ, Rabbit and SQL would not be required.  
ASQ and ASB will still need it though.

At the moment run the artifacts on the Azure VM you need to manually promote the TC build to the "Initial VMs" step.  Once we happy it's working properly we can look at adding a trigger.  

Because the VMs are being built from images the  startup time for the environment is around 10 minutes.

VM sizes can be adjusted by editing the JSON embedded in the PS Script in the "Initial VMs" step.  
Current config is using 17 of the 20 allowed cores in that subscription.  If we want to bump the configs up we'll need to request the core limit be raised via a Azure support request.

All VMs are stopped after each run to keep the cost down.  To be doubly sure I'll look at putting in place a periodic check to see how long the VMs have been running and kill them if it's over a threshold.  (A hung build would be expensive).  Current config would cost around USD $2000 per month if we ran it 24/7

**Comment by SzymonPobiega** [comment]
Just curious, do the pert tests for receiving include the auditing of the processed messages?

**Comment by WilliamBZA** [comment]
@SzymonPobiega There is a test that compares the performance with audit on/off with different configuration permutations.

http://deploy.particular.net:8000/en-US/app/search/audit_on_vs_off (password in lastpass)

**Comment by WilliamBZA** [comment]
Feedback from the retrospective:

### Things that the TF could have improved on

* Lines between vCurrent and vNext seem blurred. How do we decide what is vCurrent vs vNext?
* Decision making process around splunk wasn't interrogated well enough. We seem to have ended up with a produce that is not the best.
* Timezone mix is possibly not a good thing due to the brainstorming needed at the beginning. Having multiple timezones involved would slow this process down.

### Things that went well

* Active communication. Both in channel and zoom. Communication was very open.
* Attitude to deliver was good. Everyone was on the same page.

### Things to improve on

* Interrogate decision making process more thoroughly. More time performing demos, and analysing pros and cons
* Need rarlier alignment with terminologies and goals
* Need to promote the project and the results to the different maintainer groups to make sure they follow

**Comment by adamralph** [comment]
@WilliamBZA @hmemcpy @gbiellem @ramonsmits @dvdstelt @HEskandari is there anything from the retrospective that you feel could feed into new or current GH issues for process improvements in PlatDev?

**Comment by WilliamBZA** [comment]
@adamralph We had a few issues that I think would be specific to this particular TF. So I think perhaps the timezone learnings are about the most general. We didn't do a good enough job of distributing the work across timezones - almost like any work outside of the EU timezone was secondary work :disappointed:, but at the same time having most of the work happen in a single timezone meant easier communication and brainstorming. Tricky balance.

**Comment by weralabaj** [comment]
@WilliamBZA given our previous experience with performance testing it'll be also useful to get your thoughts on that subject specifically, e.g. do you feel we have enough expertise or maybe some training would be a good idea, how to go about it long-term, is having such ad-hoc tasks enough or is that not sufficient when forming conclusions, etc.

**Comment by WilliamBZA** [comment]
@weralabaj Good questions.

The idea of this TF was not to create a one-off performance result, but rather a framework to ensure that performance doesn't degrade in the future. The requirement for this to work though is that these tests are run and results are scrutinized for each change going forward.

There is a check in the Definition of Done to validate the performance impact of the change. This will hopefully force all future tasks to continue to value performance.

Expertise wise I think we're covered. We have a number of people that are passionate and have suitable experience.
